server/src/main/java/org/opensearch/identity/ScopeAwareSubject.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.identity;
+
+import java.util.Set;
+import org.opensearch.identity.scopes.Scope;
+
+/**
+ * This interface defines a ScopeAwareSubject.
+ *
+ * A ScopeAwareSubject is an extension of the Subject interface which adds the concept of scopes.
+ * Specifically, a ScopeAwareSubject must be able to get its own scopes (as a set of Strings) and set its own scopes,
```

Comment:
If a subject has its own scopes and they are modifiable, doesn't this mean a subject could grant itself permission to whatever it wanted?

ApplicationManager seems like the right place to understand the relationship between an Application and the scopes it is allowed.  An application aware subject would know if it was associated with an application or not, and then if trying to check if it has permission for a scope it could check at access time through the application manager if it was allowed.

It might help to make a diagram of the relationship between these different objects and their interactions with one another

Reply:
Hi @peternied, 

I made two diagrams to try to clarify things.

This first one, shows the relationship between the different Subject interfaces and the classes which implement them. There are two types of arrows. The black/solid ones show classes implementing the methods of the interface they point to. The empty/white arrows show another interface extending the interface its arrow points to. (Seems like Class Diagrams don't render on GitHub for some reason). 

![mermaid-diagram-2023-06-27-101819](https://github.com/opensearch-project/OpenSearch/assets/65832608/0fad6c00-c33b-4ec5-9273-5a6187b17845)

The second diagram shows the relationship between the ApplicationManager, the ExtensionManager, and a future PluginManager/modified PluginService. 


```mermaid
sequenceDiagram
loop Start Action
    NodeClient/RestClient->>NodeClient/RestClient: Get Subject
end
NodeClient/RestClient ->> ApplicationManager: isAllowed
loop ResolveApplication
    ApplicationManager->>ApplicationManager: Determine extension vs. plugin vs. client
end
Note right of ApplicationManager: Sub-Manager = App Type
ApplicationManager->>Sub-Manager: Fetch Application Data
Sub-Manager-->>ApplicationManager: App info
ApplicationManager-->>NodeClient/RestClient: isAllowed() response
```
The flow being demonstrated in the second diagram sees the Node or Rest Client begin the process of initiating a request based off of some transport or REST action. The client then checks the Subject and then forwards the request to the ApplicationManager with the associated subject and runs `isAllowed`. The ApplicationManager then determines if the subject is an extension, plugin, or client (an eventual way of handling users where user-initiated requests are treated as coming from an external client application). 

The manager associated with the application type can then resolve scopes for the application and return them for the ApplicationManager to be resolved. The response is then forwarded and the action allowed or disallowed.